### PR TITLE
(NETDEV-29) Enhance ntp_config and ntp_server

### PR DESCRIPTION
**Pull request by shermdog**
This PR enhances the existing ntp_config and ntp_server types as specified in  puppetlabs/netdev_stdlib#21

ntp_config
 - authenticate
 - trusted_key

ntp_server
 - key
 - maxpoll
 - minpoll
 - source_interface
 - vrf

ntp_auth_key
  - key
  - algorithm
  - mode
  - password

ntp_server initialize now uses options hash
  * Prior to this release ntp_server accepted positional arguments for id and
  prefer.  New behavior is to pass attributes as a hash.

  Example:
  ```
  options = { 'name' => id, 'key' => '999', 'prefer' => 'true',
              'minpoll' => '5', 'maxpoll' => '8', 'vrf' => 'red' }
  Cisco::NtpServer.new(options, true)
  ```

### CODE REVIEW

**Comment by shermdog** [comment]
Opening PR to review current methods, code, etc.  Will be adding tests shortly.

File: lib/cisco_node_utils/ntp_server.rb [hunk 1]
```diff
@@ -66,20 +64,41 @@ def ==(other)
...
       config_set('ntp_server', 'server', state: '', ip: @ntpserver_id,
-                  prefer: @ntpserver_prefer ? 'prefer' : '')
+                  prefer: @prefer == 'true' ? 'prefer' : '',
+                  vrf: @vrf ? "use-vrf #{@vrf}" : '',
+                  key: @key ? "key #{@key}" : '',
+                  minpoll: @minpoll ? "minpoll #{@minpoll}" : '',
+                  maxpoll: @maxpoll ? "maxpoll #{@maxpoll}" : '')
     end
 
     def destroy
```
Comment (mikewiebe) [inline lib/cisco_node_utils/ntp_server.rb:76]: I believe you may be able to simplify the `destroy` method and only specify the `@ntpserver_id`. I did some quick testing and I was able to remove the config without specifying any of the other parameters.

`config_set('ntp_server', state: 'no', ip: @ntpserver_id)`

File: lib/cisco_node_utils/ntp_server.rb [hunk 2]
```diff
@@ -66,20 +64,41 @@ def ==(other)
...
-                 state: 'no', ip: @ntpserver_id, prefer: '')
+                 state: 'no', ip: @ntpserver_id, prefer: '', vrf: '',
+                 key: '', minpoll: '', maxpoll: '')
     end
 
     def name
       @ntpserver_id
     end
 
+    def key
```
Comment (mikewiebe) [inline lib/cisco_node_utils/ntp_server.rb]: The following is simpler and will get rid of the rubocop errors for the object properties.

`attr_reader :key, :maxpoll, :minpoll, :prefer, :vrf`

**Comment by mikewiebe** [comment]
@shermdog Looks good to me.  I gave you a few inline comments and here are a few general housekeeping things:

* Update Copyright date ranges for any files you touch
* Update the https://github.com/cisco/cisco-network-node-utils/blob/develop/CHANGELOG.md#changelog
* Update/add tests in:
  * https://github.com/cisco/cisco-network-node-utils/blob/develop/tests/test_ntp_config.rb
  * https://github.com/cisco/cisco-network-node-utils/blob/develop/tests/test_ntp_server.rb
* Resolve the rubocop errors.

File: lib/cisco_node_utils/ntp_server.rb (hunk 1 above)
Comment (shermdog) [reply lib/cisco_node_utils/ntp_server.rb:76]: I get an exception without it

```
No value specified for 'prefer' in 'no ntp server 2.2.2.2 <prefer> <vrf> <key> <minpoll> <maxpoll>'
```

Comment (mikewiebe) [reply lib/cisco_node_utils/ntp_server.rb:76]: Ah, that's because the other properties need to get passed in as empty string.  I suppose this is fine.

File: tests/test_ntp_config.rb [hunk 3]
```diff
@@ -26,28 +26,75 @@ def setup
...
+    # teardown runs at the end of each test
+    no_ntpconfig
+    super
+  end
+
   def no_ntpconfig
     # Turn the feature off for a clean test.
     if platform == :ios_xr
-      config("no ntp source #{interfaces[0]}")
+      config("no ntp source #{interfaces[1]}")
```
Comment (mikewiebe) [inline tests/test_ntp_config.rb:38]: Why the change from `interface[0]` to `interface[1]`?

Comment (shermdog) [reply tests/test_ntp_config.rb:38]: The test was using `interface[1]` but the teardown was not.  This was leaving cruft behind after the tests were complete.

File: tests/test_ntp_server.rb [hunk 4]
```diff
@@ -83,4 +92,54 @@ def test_multiple
     refute_includes(Cisco::NtpServer.ntpservers, id1)
     refute_includes(Cisco::NtpServer.ntpservers, id2)
   end
+
+  def test_domain_name
+    id = '0.us.pool.ntp.org'
```
Comment (mikewiebe) [inline tests/test_ntp_server.rb]: When I run this on my testbed it fails because we are behind a proxy.  This is fine for now, but I will probably move this into a yaml configuration file that automatically skips the test if the config file is unavailable and then anyone that wants to run this test can add any internal or external servers

File: tests/test_ntp_server.rb [hunk 5]
```diff
@@ -83,4 +92,54 @@ def test_multiple
     refute_includes(Cisco::NtpServer.ntpservers, id1)
     refute_includes(Cisco::NtpServer.ntpservers, id2)
   end
+
+  def test_domain_name
```
Comment (mikewiebe) [inline tests/test_ntp_server.rb]: Does `test_domain_name` pass for you?  When I run it on my testbed the configuration is accepted but the ntp server does not actually get configured...  investigating to see if it's just the version of software I am running.

File: tests/test_ntp_config.rb (hunk 3 above)
Comment (mikewiebe) [reply tests/test_ntp_config.rb:38]: 👍

File: tests/test_ntp_server.rb (hunk 5 above)
Comment (shermdog) [reply tests/test_ntp_server.rb]: It does, but requires DNS to be functional in the default VRF.

On some platforms / os the device will silently fail if it cannot resolve the hostname.

I'm actually okay with taking out the hostname test as it's problematic.

**Comment by mikewiebe** [comment]
@shermdog I want to wait on merging this until we are ready to merge https://github.com/cisco/cisco-network-puppet-module/pull/447.  If we merge now, our beaker tests will start to fail because of the changes you made to pass opts as a hash instead of positional args.

File: tests/test_ntp_server.rb (hunk 5 above)
Comment (mikewiebe) [reply tests/test_ntp_server.rb]: Yeah.. let's just remove it for now.

**Comment by mikewiebe** [comment]
👍
